### PR TITLE
add runtime env manager

### PR DESCRIPTION
**Pull request by runzexia**

### CODE REVIEW

**Comment by zheng1** [comment]
I refactor `db.go`, and the diff is following. The test case will generate SQL like that. Please compare whether it is same with your expect.
```sql
SELECT * FROM runtime_env 
JOIN `runtime_env_label` AS `table_label_0` ON table_label_0.runtime_env_id = runtime_env.runtime_env_id 
JOIN `runtime_env_label` AS `table_label_1` ON table_label_1.runtime_env_id = runtime_env.runtime_env_id 
WHERE (
    (
        (`table_label_0`.`label_key` = 'env') AND 
        (`table_label_0`.`label_value` IN ('testing'))
    ) AND 
    (
        (`table_label_1`.`label_key` = 'runtime') AND 
        (`table_label_1`.`label_value` IN ('k8s','qingcloud'))
    )
)
```

```diff
diff --git a/pkg/manager/runtime_env/db.go b/pkg/manager/runtime_env/db.go
index 294a1c0..584dd43 100644
--- a/pkg/manager/runtime_env/db.go
+++ b/pkg/manager/runtime_env/db.go
@@ -301,6 +301,36 @@ func BuildDeleteLabelFilterConditions(runtimeEnvId, labelKey, labelValue string)
 	return db.And(conditions...)
 }
 
+type Table interface {
+	GetTableName() string
+	GetPrimaryKey() string
+}
+
+type TableHadLabel interface {
+	Table
+	GetLabelTableName() string
+}
+
+func BuildLabelsStmt(selectorMap map[string][]string, table TableHadLabel) *dbr.SelectStmt {
+	baseTableName := table.GetTableName()
+	labelTabelName := table.GetLabelTableName()
+	primaryKey := table.GetPrimaryKey()
+	q := dbr.Select("*").From(baseTableName)
+	i := 0
+	var whereCondition []dbr.Builder
+	for key, values := range selectorMap {
+		aliasTableName := fmt.Sprintf("table_label_%d", i)
+		onCondition := fmt.Sprintf("%s.%s = %s.%s", aliasTableName, primaryKey, baseTableName, primaryKey)
+		q = q.Join(dbr.I(labelTabelName).As(aliasTableName), onCondition)
+		whereCondition = append(whereCondition, db.And(db.Eq(aliasTableName+".label_key", key), db.Eq(aliasTableName+".label_value", values)))
+		i++
+	}
+	if len(whereCondition) > 0 {
+		q = q.Where(db.And(whereCondition...))
+	}
+	return q
+}
+
 func BuildSelectRuntimeEnvIdBySelectorSql(selectorMap map[string][]string) string {
 	i := 0
 	var selectConditionArray []string
diff --git a/pkg/manager/runtime_env/db_test.go b/pkg/manager/runtime_env/db_test.go
new file mode 100644
index 0000000..ca389eb
--- /dev/null
+++ b/pkg/manager/runtime_env/db_test.go
@@ -0,0 +1,25 @@
+// Copyright 2018 The OpenPitrix Authors. All rights reserved.
+// Use of this source code is governed by a Apache license
+// that can be found in the LICENSE file.
+
+package runtime_env
+
+import (
+	"testing"
+
+	"github.com/gocraft/dbr"
+	"github.com/gocraft/dbr/dialect"
+	"github.com/stretchr/testify/assert"
+
+	"openpitrix.io/openpitrix/pkg/models"
+)
+
+func TestBuildLabelsStmt(t *testing.T) {
+	q := BuildLabelsStmt(map[string][]string{"runtime": {"k8s", "qingcloud"}, "env": {"testing"}}, models.RuntimeEnv{})
+	buf := dbr.NewBuffer()
+	err := q.Build(dialect.MySQL, buf)
+	assert.NoError(t, err)
+	s, err := dbr.InterpolateForDialect(buf.String(), buf.Value(), dialect.MySQL)
+	assert.NoError(t, err)
+	t.Logf(s)
+}
diff --git a/pkg/models/runtime_env.go b/pkg/models/runtime_env.go
index 8cc7afd..37e072e 100644
--- a/pkg/models/runtime_env.go
+++ b/pkg/models/runtime_env.go
@@ -29,6 +29,12 @@ type RuntimeEnv struct {
 	StatusTime    time.Time
 }
 
+func (_ RuntimeEnv) GetTableName() string { return RuntimeEnvTableName }
+
+func (_ RuntimeEnv) GetPrimaryKey() string { return "runtime_env_id" }
+
+func (_ RuntimeEnv) GetLabelTableName() string { return RuntimeEnvLabelTableName }
+
 var RuntimeEnvColumns = GetColumnsFromStruct(&RuntimeEnv{})
 
 func NewRuntimeEnv(name, description, runtimeEnvUrl, owner string) *RuntimeEnv {

```

**Comment by runzexia** [comment]
It's basically consistent with my except, but there are a few questions.
Should not go to select * from table, because this will query all fields, including the fields of the table that joins.
The `runtime_env_id` is not PrimaryKey,It might be better to name JoinKey.

In the code I wrote, I spliced my own SQL without using dbr.Because I don't know how DBR's alias is used.After reading your code I learned how to use it and I'm going to test the correctness of the query.

**Comment by runzexia** [comment]
It's basically consistent with my except, but there are a question.
Should not go to select * from table, because this will query all fields, including the fields of the table that joins.

In the code I wrote, I spliced my own SQL without using dbr.Because I don't know how DBR's alias is used.After reading your code I learned how to use it and I'm going to test the correctness of the query.

File: pkg/manager/runtime_env/check.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+package runtime_env
+
+import (
+	"fmt"
+
+	"openpitrix.io/openpitrix/pkg/constants"
+	"openpitrix.io/openpitrix/pkg/logger"
+)
+
+func (p *Server) checkRuntimeEnvDeleted(runtimeEnvId string) (bool, error) {
```
Comment (zheng1) [inline pkg/manager/runtime_env/check.go]: If the func is not reusability, it should write to the top function.

File: pkg/manager/runtime_env/handler.go [hunk 2]
```diff
@@ -8,40 +8,631 @@ import (
...
+	res := &pb.DeleteRuntimeEnvResponse{
+		RuntimeEnv: pbRuntimeEnv,
+	}
+	return res, nil
 }
+
 func (p *Server) CreateRuntimeEnvCredential(ctx context.Context, req *pb.CreateRuntimeEnvCredentialRequset) (*pb.CreateRuntimeEnvCredentialResponse, error) {
-	return nil, status.Errorf(codes.Internal, "CreateRuntimeEnvCredential: %+v", fmt.Errorf("hello world"))
+	s := sender.GetSenderFromContext(ctx)
+	logger.Debugf("Got sender: %+v", s)
```
Comment (zheng1) [inline pkg/manager/runtime_env/handler.go]: Remove this debug log, we will add it to grpc level.
```go
logger.Debugf("Got sender: %+v", s)
logger.Debugf("Got req: %+v", req)
```

File: pkg/manager/runtime_env/handler.go [hunk 3]
```diff
@@ -8,40 +8,631 @@ import (
...
+		return nil, status.Errorf(codes.Internal, "AttachCredentialToRuntimeEnv: %+v", err)
+	}
+	if attachment != nil {
+		logger.Errorf("runtime_env has been attached to %s", attachment.RuntimeEnvCredentialId)
+		return nil, status.Errorf(codes.Internal,
+			"AttachCredentialToRuntimeEnv: runtime_env has been attached to %s",
+			attachment.RuntimeEnvCredentialId)
+	}
+
+	err = p.createRuntimeEnvAttachedCredential(runtimeEnvId, runtimeEnvCredentialId)
```
Comment (zheng1) [inline pkg/manager/runtime_env/handler.go]: This func should be named with `attachCredentialToRuntimeEnv `

File: pkg/manager/runtime_env/handler.go [hunk 4]
```diff
@@ -8,40 +8,631 @@ import (
...
+		return nil, status.Errorf(codes.Internal,
+			"DetachCredentialFromRuntimeEnv: failed to get runtime_env_attached_credential [%v]", err)
+	}
+	runtimeEnvCredentialId := req.RuntimeEnvCredentialId.GetValue()
+	if attachment.RuntimeEnvCredentialId != runtimeEnvCredentialId {
+		logger.Errorf("runtime_env_credential value not match")
+		return nil, status.Errorf(codes.Internal,
+			"DetachCredentialFromRuntimeEnv: runtime_env_credential value not match")
+	}
+	err = p.deleteRuntimeEnvAttachment(runtimeEnvId, runtimeEnvCredentialId)
```
Comment (zheng1) [inline pkg/manager/runtime_env/handler.go]: This func should be named with `detachCredentialFromRuntimeEnv`

**Pull request merged**